tables/TaQL/ExprNodeSetOpt.cc

```diff
@@ -0,0 +1,450 @@
+    Vector<Int64> index;
+    GenSortIndirect<T,Int64>::sort (index, stvals, stvals.size());
+    std::vector<T> newStart;
+    std::vector<T> newEnd;
+    std::vector<Bool> newLeftC;
+    std::vector<Bool> newRightC;
+    std::vector<rownr_t> rowNrs;
+    if (!combine) {
+      for (size_t i=0; i<index.size(); ++i) {
+        Int64 inx = index[i];
+        newStart.push_back (stvals[inx]);
+        newEnd.push_back (endvals[inx]);
+        newLeftC.push_back (set[inx]->isLeftClosed());
+        newRightC.push_back (set[inx]->isRightClosed());
+        rowNrs.push_back (inx);
+      }
+    } else {
+      // Get the start and end value of first interval in sorted list.
+      T stval  = stvals[index[0]];
+      T endval = endvals[index[0]];
+      Bool leftC  = set[index[0]]->isLeftClosed();
+      Bool rightC = set[index[0]]->isRightClosed();
+      rownr_t rownr = index[0];
+      // Loop through the next intervals and combine if possible.
+      for (size_t i=1; i<index.size(); i++) {
+        Int64 inx = index[i];
+        T st2 = stvals[inx];
+        T end2 = endvals[inx];
+        // Combine intervals if they overlap.
+        // They do if the next interval starts before end of this one
+        // or if starting at the end and one side of the interval is closed.
+        if (st2 < endval  ||
+            (st2 == endval  &&
+             (set[inx]->isLeftClosed() || set[index[i-1]]->isRightClosed()))) {
+          // Overlap; update end if higher.
+          if (end2 > endval) {
+            endval = end2;
+            rightC = set[inx]->isRightClosed();
+          }
+        } else {
+          // No overlap, so create the interval found and start a new one.
+          newStart.push_back (stval);
+          newEnd.push_back (endval);
+          newLeftC.push_back (leftC);
+          newRightC.push_back (rightC);
+          rowNrs.push_back (rownr);
+          stval  = st2;
+          endval = end2;
+          leftC  = set[inx]->isLeftClosed();
+          rightC = set[inx]->isRightClosed();
+          rownr  = inx;
+        }
+      }
+      // Create the last interval.
+      newStart.push_back (stval);
+      newEnd.push_back (endval);
+      newLeftC.push_back (leftC);
+      newRightC.push_back (rightC);
+      rowNrs.push_back (rownr);
+    }
+    // Now create the correct object.
+    return createOptSet (set, newStart, newEnd, newLeftC, newRightC);
+  }
+
+  template<typename T>
+  TENShPtr TableExprNodeSetOptContSet<T>::createOptSet
+  (const TableExprNodeSet& set,
+   const std::vector<T>& start, const std::vector<T>& end, 
+   const std::vector<Bool>& leftC, const std::vector<Bool>& rightC)
+   {
+    // See if all intervals have the same left and right closedness.
+    // If so, a better version can be used that does not need to test on it
+    // for each compare.
+    Bool same = True;
+    for (size_t i=1; i<leftC.size(); ++i) {
+      if (leftC[i] != leftC[0]  ||  rightC[i] != rightC[0]) {
```

Comment:
There's no constraint that `leftC.size() == rightC.size()`

Reply:
Added